aten/src/ATen/TensorIterator.cpp

```diff
@@ -343,14 +348,18 @@ void TensorIteratorBase::compute_types(const TensorIteratorConfig& config) {
     return;
   }
 
+  common_dtype_ = config.specified_common_dtype.value_or(common_dtype_);
+
   // Computes a common dtype, if needed
-  if (has_different_input_dtypes && config.promote_inputs_to_common_dtype_) {
+  if (has_different_input_dtypes && config.promote_inputs_to_common_dtype_ &&
```

Comment:
cc @ngimel 

TensorIterator is one of our most significant performance pain points, so we need to be very careful about how we modify it. 

I'm not especially concerned with the perf AFTER we're in the branch where a user has specified a common dtype, so the trick is going to be minimizing the time it takes to get to the appropriate branch. Can we do this without going too crazy on the code? I think so, because I think we can edit the initialization of common_dtype_ on line 268 above to be config.specified_common_dtype. Instead of representing that value as an optional it can just have a default value of ScalarType::Undefined. This is a load from a memory location instead of loading a constant, but that's probably the most minimal perf hit we can take without seriously revising how we're dispatching.

So what does this do? Well, if any inputs are different from the specified common dtype it will trigger the has_different_input_dtypes flag and we'll drop into this conditional:

```
if (has_different_input_dtypes && config.promote_inputs_to_common_dtype_) {
```

This is good because we're already limiting a perf impact to a slow path, and we're (barely) increasing the time to get to that slow path.

We can then update what happens in this conditional:

```
common_dtype_ = compute_common_dtype();
if (config.specified_common_dtype != ScalarType::Undefined) {
  if (config.promote_integer_inputs_to_float_ 
      && c10::isIntegralType(config.specified_common_dtype, /*include_bool=*/true) {
    // TODO: throw an error that the operation doesn't compute with integer types, but the common dtype is integral
  }
  // TODO: check can_cast from common_dtype_ to config.specified_common_dtype here!
  common_dtype_ = config.specified_common_dtype;
}
```

This folds the check in the following conditional into the slower path for config.specified_common_dtype. So, again, the checks on the fast path are UNCHANGED. If type promotion has to occur, however, then there is an additional check for whether the user specified the dtype. I think this is OK, but let's see what @ngimel thinks. 

What do you think of an approach like this @kshitij12345?